bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,549 @@
+                    .getBundleContext().registerService(DiscoveryService.class, this.discovery, null);
+            this.discovery.activate(null);
+
+            // kick off the bridge connection process
+            this.scheduler.execute(this::connect);
+        }
+    }
+
+    private boolean validConfiguration(@Nullable SomfyMyLinkConfiguration config) {
+        if (config == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "mylink configuration missing");
+            return false;
+        }
+
+        if (StringUtils.isEmpty(config.ipAddress)) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "mylink address not specified");
+            return false;
+        }
+
+        return true;
+    }
+
+    private void connect() {
+        try {
+            SomfyMyLinkConfiguration config = this.config;
+
+            if (config == null) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "mylink config not specified");
+                return;
+            }
+
+            if (StringUtils.isEmpty(config.ipAddress) || StringUtils.isEmpty(config.systemId)) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "mylink config not specified");
+                return;
+            }
+
+            // start the keepalive process
+            ensureKeepAlive();
+
+            logger.debug("Connecting to mylink at {}", config.ipAddress);
+
+            // send a ping
+            sendPing();
+
+            logger.debug("Connected to mylink at {}", config.ipAddress);
+
+            updateStatus(ThingStatus.ONLINE);
+        } catch (SomfyMyLinkException e) {
+            logger.debug("Problem connecting to mylink, bridge OFFLINE");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
+    }
+
+    private void ensureKeepAlive() {
+        if (heartbeat == null) {
+            logger.debug("Starting keepalive job in {} min, every {} min", HEARTBEAT_MINUTES, HEARTBEAT_MINUTES);
+            heartbeat = this.scheduler.scheduleWithFixedDelay(this::sendKeepAlive, 1, 1, TimeUnit.MINUTES);
+        }
+    }
+
+    private void disconnect() {
+        logger.debug("Disconnecting from mylink");
+        ScheduledFuture<?> heartbeat = this.heartbeat;
+
+        if (heartbeat != null) {
+            logger.debug("Cancelling keepalive job");
+            heartbeat.cancel(true);
+            heartbeat = null;
+        } else {
+            logger.debug("Keepalive was not active");
+        }
+    }
+
+    private void sendKeepAlive() {
+        try {
+            logger.debug("Keep alive triggered");
+
+            if (getThing().getStatus() != ThingStatus.ONLINE) {
+                // try connecting
+                logger.debug("Bridge offline, trying to connect");
+                connect();
```

Comment:
Why is this getting called rather than just setting the thing to online if the ping succeeds?
I think it adds a bit of confusion having `connect` and `sendKeepAlive` both responsible for updating the thing status. 
To add to the confusion it doesn't look like `connect` or `disconnect` have anything to do with connecting or disconnecting. `connect` simply starts the keepalive process and then sends a ping. `disconnect` just stops the keepalive process. You should rename these methods to something more appropriate.

I think it might be better to get rid of the `connect` method entirely and make `sendKeepAlive` responsible for changing the thing status to online/offline. `disconnect` should probably be renamed `cancelHeartbeat`.

Reply:
yes this was really confusing. vastly simplified now.